tony-core/src/main/java/com/linkedin/tony/runtime/MLGenericRuntime.java

```diff
@@ -142,12 +188,85 @@ protected void printTasksPeriodically() {
     // ===================For Task Executor=======================
 
     @Override
-    public int run(TaskExecutor executor) throws Exception {
-        buildTaskEnv(executor);
-        return executorPythonShell(executor);
+    public void initTaskExecutorResource(TaskExecutor executor) {
+        this.taskExecutor = executor;
     }
 
-    protected void buildTaskEnv(TaskExecutor executor) throws Exception {
-        return;
+    /**
+     * The tensorboard port will only be reserved on chief or sidecar tensorboard executor.
+     * Other executors will not reserve it.
+     * However, the above reserved port action will not be at the same time.
+     * If enable side-car tensorboard, port will only be reserved on it, otherwise will only be on chief executor.
+     */
+    @Override
+    public boolean needReserveTBPort() {
+        assert taskExecutor != null;
+
+        boolean enableSidecarTB = enableSidecarTB(taskExecutor.getTonyConf());
+
+        // When disable sidecar tensorboard, it will only be reserved on chief task executor.
+        if (!enableSidecarTB && taskExecutor.isChief()) {
+            return true;
+        }
+
+        // When enable sidecar tensorboard, it will only be reserved on sidecar executor.
+        if (enableSidecarTB && isSidecarTBExecutor(taskExecutor)) {
+            return true;
+        }
+
+        return false;
     }
+
+    @Override
+    public int run() throws Exception {
+        assert taskExecutor != null;
+
+        if (isSidecarTBExecutor(taskExecutor)) {
+            // inject starting tensorboard command
+            setStartupTBResource();
+        }
+
+        buildTaskEnv(taskExecutor);
+        return executorPythonShell(taskExecutor);
+    }
+
+    private void setStartupTBResource() {
+        String tbLogDir = taskExecutor.getTonyConf().get(TonyConfigurationKeys.TENSORBOARD_LOG_DIR);
+        Path tbScriptPath = createSidecarTBScript();
+        String pythonExecPath = taskExecutor.getTonyConf().get(TonyConfigurationKeys.PYTHON_EXEC_PATH, "python");
+        String startTBCommand = String.format("%s %s --logdir %s --port %s",
+                pythonExecPath, tbScriptPath, tbLogDir, taskExecutor.getTbPort());
```

Comment:
how is this registered with AM trackingURL?

Reply:
Here. https://github.com/linkedin/TonY/blob/9010e0272923a51698954139bab68d47c1dcf69d/tony-core/src/main/java/com/linkedin/tony/TaskExecutor.java#L98